### PR TITLE
Blocks

### PR DESCRIPTION
**Pull request by mastercoms**
A rewrite of a lot of things in net.glowstone.blocks

I removed the Sponge stuff already, so don't worry about that stuff getting in.

Very WIP. Lots of work to still be done. Two major things that are pending are getting blocks to be single instance per Location, and rewriting the blocktype package to allow for better inheritance and custom blocks.

### CODE REVIEW

File: src/main/java/net/glowstone/block/GlowBlock.java [hunk 1]
```diff
@@ -212,48 +211,101 @@ public void setType(Material type, byte data, boolean applyPhysics) {
...
+            getChunk().setType(x & 0xf, z & 0xf, y, type);
+            this.type = type;
+            changed = true;
+        }
+        if (oldData != data) {
+            getChunk().setMetaData(x & 0xf, z & 0xf, y, data);
+            changed = true;
+        }
 
-        if (oldTypeId == Material.DOUBLE_PLANT && getRelative(BlockFace.UP).getType() == Material.DOUBLE_PLANT) {
```
Comment (mastercoms) [inline src/main/java/net/glowstone/block/GlowBlock.java:233]: This needs to be moved to the blocktype stuff once that's rewritten.

**Comment by mastercoms** [comment]
I am also probably going to redo the work from SpaceManiac, rewritten to use Bukkit instead of Sponge, with some other enhancements in light of changes in Glowstone and the recent changes to the blocks branch.

File: src/main/java/net/glowstone/block/MaterialValueManager.java [hunk 2]
```diff
@@ -98,15 +423,29 @@ public GlowMaterialBuilder setRandomTicks(boolean randomTicks) {
...
+                return pistonMoveReaction;
+            }
+
+            public GlowMaterialBuilder setPistonMoveReaction(PistonMoveReaction pistonMoveReaction) {
+                this.pistonMoveReaction = pistonMoveReaction;
+                return this;
+            }
+        }
+
+        GlowMaterial() {
```
Comment (mastercoms) [inline src/main/java/net/glowstone/block/MaterialValueManager.java:568]: This is a temporary constructor until all values are filled in.

File: src/main/java/net/glowstone/block/GlowBlock.java [hunk 3]
```diff
@@ -212,48 +211,101 @@ public void setType(Material type, byte data, boolean applyPhysics) {
...
+
     @Override
     public boolean setTypeId(int type) {
         return setTypeId(type, true);
     }
 
     @Override
     public boolean setTypeId(int type, boolean applyPhysics) {
-        return setTypeIdAndData(type, (byte) 0, applyPhysics);
+        return setTypeIdAndData(type, getData(), applyPhysics);
```
Comment (mastercoms) [inline src/main/java/net/glowstone/block/GlowBlock.java]: The spec only tells us to set the type, but should we reset the data as well?

Comment (aramperes) [reply src/main/java/net/glowstone/block/GlowBlock.java]: As to keep the same functionality as Craftbukkit/Spigot, it should get reset to the default data value of the new block

**Comment by gdude2002** [comment]
This has been here for a while - it also has some conflicts. Do we need to keep this PR going?

**Comment by mastercoms** [comment]
Yes, I will get around to updating it soon, after I get the new chat stuff working.

**Comment by mastercoms** [comment]
Closing this as a new PR is coming soon.

**Pull request closed without merging**